File: Sdk/DotNet/bNesis.Sdk.Common.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Net;

namespace bNesis.Sdk.Common
{
	///<summary>
	/// Information about error 
	/// </summary>
	public class ErrorInfo
	{
		/// <summary>
		/// Error code 
		/// </summary>
		public Int32 Code { get; set; }

		/// <summary>
		/// When ErrorInfo was created 
		/// </summary>
		public string DateTime { get; set; }

		/// <summary>
		/// Service name 
		/// </summary>
		public string Service { get; set; }

		/// <summary>
		/// Describes important error information 
		/// </summary>
		public string Description { get; set; }

		/// <summary>
		/// All information about error 
		/// </summary>
		public string BasicDescription { get; set; }

	}

	///<summary>
	/// A generic class that defines the HTTP result of the API call in bNesse
	///     Each bNesis services RAW API must return this class (used as data structure) to Client 
	/// </summary>
	public class Response
	{
		/// <summary>
		/// HTTP response status code 
		/// </summary>
		public HttpStatusCode StatusCode { get; set; }

		/// <summary>
		/// HTTP response status code text description 
		/// </summary>
		public string StatusDescription { get; set; }

		/// <summary>
		/// String representation of response content 
		/// </summary>
		public string Content { get; set; }

		/// <summary>
		/// Encoding of the response content 
		/// </summary>
		public string ContentEncoding { get; set; }

		/// <summary>
		/// Length in bytes of the response content 
		/// </summary>
		public Int64 ContentLength { get; set; }

		/// <summary>
		/// MIME content type of response 
		/// </summary>
		public string ContentType { get; set; }

		/// <summary>
		/// Bytes 
		/// </summary>
		public Byte[] RawBytes { get; set; }

		/// <summary>
		/// Headers at JSON temporary solution 
		/// </summary>
		public string Headers { get; set; }

		/// <summary>
		/// message with error information 
		/// </summary>
		public string ErrorMessage { get; set; }

		/// <summary>
		/// Status of the request. Will return Error for transport errors. HTTP errors will
		/// still return ResponseStatus.Completed, check StatusCode instead 
		/// </summary>
		public string ResponseStatus { get; set; }

		/// <summary>
		/// The URL that actually responded to the content (different from request if redirected) 
		/// </summary>
		public Uri ResponseUri { get; set; }

		/// <summary>
		/// HttpWebResponse.Server 
		/// </summary>
		public string Server { get; set; }

	}

	///<summary>
	/// Only for testing class at class 
	/// </summary>
	public class AddressTest
	{
		/// <summary>
		/// some Country simple property 
		/// </summary>
		public string Country { get; set; }

		/// <summary>
		/// some City simple string property 
		/// </summary>
		public string City { get; set; }

		/// <summary>
		/// City streets array for test - at Class AddressTest exists complex property Street[] - is array of strings 
		/// </summary>
		public string[] Street { get; set; }

	}

	///<summary>
	/// bNesis unified contact item for services unified API call 
	/// </summary>
	public class ContactItem
	{
		/// <summary>
		/// contact address at free format - just string 
		/// </summary>
		public string Address { get; set; }

		/// <summary>
		/// contact email, is class is not check are the email typed at right email format 
		/// </summary>
		public string Email { get; set; }

		/// <summary>
		/// contact data block (record) Id - is unical for all service contact 
		/// </summary>
		public string Id { get; set; }

		/// <summary>
		/// this is test property for formalised Address - street, home, post code 
		/// </summary>
		public AddressTest AddressTestProp { get; set; }

		/// <summary>
		/// Contact name - it maybe single name or FNAMEspaceLNAME 
		/// </summary>
		public string Name { get; set; }

		/// <summary>
		/// contact phone at free format, this class is not checkid it 
		/// </summary>
		public string Phone { get; set; }

	}

	///<summary>
	/// unified bNesis product variants defenition class 
	/// </summary>
	public class Variants
	{
		/// <summary>
		/// product variant options 
		/// </summary>
		public string Option { get; set; }

		/// <summary>
		/// product price variant 
		/// </summary>
		public string Price { get; set; }

		/// <summary>
		/// product sku variant 
		/// </summary>
		public string Sku { get; set; }

	}

	///<summary>
	/// bNesis unified product item class for services unified API call 
	/// </summary>
	public class ProductItem
	{
		public string Title { get; set; }

		public string BodyHtml { get; set; }

		public string Vendor { get; set; }

		public string ProductType { get; set; }

		public Variants[] Variants { get; set; }

	}

	///<summary>
	/// bNesis unified Coutry item for services unified API call 
	/// </summary>
	public class CountryItem
	{
		/// <summary>
		/// Country data block (record) unical Id 
		/// </summary>
		public string Id { get; set; }

		/// <summary>
		/// The Name of the Country 
		/// </summary>
		public string Name { get; set; }

		/// <summary>
		/// Coutry Code
		/// https://en.wikipedia.org/wiki/Country_code
		/// ISO 
		/// </summary>
		public string Code { get; set; }

		/// <summary>
		/// https://en.wikipedia.org/wiki/List_of_countries_by_tax_rates 
		/// </summary>
		public string Tax { get; set; }

	}

	///<summary>
	/// bNesis balance class for unified API 
	/// </summary>
	public class Balance
	{
		/// <summary>
		/// currency type flag 
		/// </summary>
		public string Currency { get; set; }

		/// <summary>
		/// Currency ammount value 
		/// </summary>
		public Double Amount { get; set; }

	}

	///<summary>
	/// bNesis balance result unified class 
	/// </summary>
	public class RetrieveBalanceUnified
	{
		/// <summary>
		/// this record Id - define data block at service 
		/// </summary>
		public string Id { get; set; }

		/// <summary>
		/// Amount value 
		/// </summary>
		public Double Amount { get; set; }

		/// <summary>
		/// Currency type flag 
		/// </summary>
		public string Currency { get; set; }

		/// <summary>
		/// data block (record) creation date time 
		/// </summary>
		public string Created { get; set; }

	}

	public class PaymentsArchiveData
	{
		public Int64 PaymentId { get; set; }

		public Double Amount { get; set; }

		public string Currency { get; set; }

		public Int64 CreateDate { get; set; }

		public Int64 TransactionId { get; set; }

	}

	public class PaymentsArchive
	{
		public PaymentsArchiveData[] data { get; set; }

	}

}

File: Examples/Generics/DiscountCalculation/Scripts/bNesis.Sdk.Common.js
/**
 * A generic class that defines the HTTP result of the API call in bNesse
 *     Each bNesis services RAW API must return this class (used as data structure) to Client 
 * @typedef {Object} Response
 */
 Response = function () { 
	/**
	 * HTTP response status code
	 * @type {HttpStatusCode}
	 */
	this.StatusCode = new HttpStatusCode();

	/**
	 * HTTP response status code text description
	 * @type {string}
	 */
	this.StatusDescription = "";

	/**
	 * String representation of response content
	 * @type {string}
	 */
	this.Content = "";

	/**
	 * Encoding of the response content
	 * @type {string}
	 */
	this.ContentEncoding = "";

	/**
	 * Length in bytes of the response content
	 * @type {Int64}
	 */
	this.ContentLength = 0;

	/**
	 * MIME content type of response
	 * @type {string}
	 */
	this.ContentType = "";

	/**
	 * Bytes
	 * @type {Byte[]}
	 */
	this.RawBytes = new Array();

	/**
	 * Headers at JSON temporary solution
	 * @type {string}
	 */
	this.Headers = "";

	/**
	 * message with error information
	 * @type {string}
	 */
	this.ErrorMessage = "";

	/**
	 * Status of the request. Will return Error for transport errors. HTTP errors will
	 * still return ResponseStatus.Completed, check StatusCode instead
	 * @type {string}
	 */
	this.ResponseStatus = "";

	/**
	 * The URL that actually responded to the content (different from request if redirected)
	 * @type {Uri}
	 */
	this.ResponseUri = new Uri();

	/**
	 * HttpWebResponse.Server
	 * @type {string}
	 */
	this.Server = "";

}

/**
 * Information about error 
 * @typedef {Object} ErrorInfo
 */
 ErrorInfo = function () { 
	/**
	 * Error code
	 * @type {Int32}
	 */
	this.Code = 0;

	/**
	 * When ErrorInfo was created
	 * @type {string}
	 */
	this.DateTime = "";

	/**
	 * Service name
	 * @type {string}
	 */
	this.Service = "";

	/**
	 * Describes important error information
	 * @type {string}
	 */
	this.Description = "";

	/**
	 * All information about error
	 * @type {string}
	 */
	this.BasicDescription = "";

}

/**
 * Only for testing class at class 
 * @typedef {Object} AddressTest
 */
 AddressTest = function () { 
	/**
	 * some Country simple property
	 * @type {string}
	 */
	this.Country = "";

	/**
	 * some City simple string property
	 * @type {string}
	 */
	this.City = "";

	/**
	 * City streets array for test - at Class AddressTest exists complex property Street[] - is array of strings
	 * @type {string[]}
	 */
	this.Street = new Array();

}

/**
 * bNesis unified contact item for services unified API call 
 * @typedef {Object} ContactItem
 */
 ContactItem = function () { 
	/**
	 * contact address at free format - just string
	 * @type {string}
	 */
	this.Address = "";

	/**
	 * contact email, is class is not check are the email typed at right email format
	 * @type {string}
	 */
	this.Email = "";

	/**
	 * contact data block (record) Id - is unical for all service contact
	 * @type {string}
	 */
	this.Id = "";

	/**
	 * this is test property for formalised Address - street, home, post code
	 * @type {AddressTest}
	 */
	this.AddressTestProp = new AddressTest();

	/**
	 * Contact name - it maybe single name or FNAMEspaceLNAME
	 * @type {string}
	 */
	this.Name = "";

	/**
	 * contact phone at free format, this class is not checkid it
	 * @type {string}
	 */
	this.Phone = "";

	/**
	 * additional information
	 * @type {string}
	 */
	this.About = "";

	/**
	 * for dynamic response
	 * @type {string}
	 */
	this.FieldsDynamic = "";

}

/**
 * unified bNesis product variants defenition class 
 * @typedef {Object} Variants
 */
 Variants = function () { 
	/**
	 * product variant options
	 * @type {string}
	 */
	this.Option = "";

	/**
	 * product price variant
	 * @type {string}
	 */
	this.Price = "";

	/**
	 * product sku variant
	 * @type {string}
	 */
	this.Sku = "";

}

/**
 * bNesis unified product item class for services unified API call 
 * @typedef {Object} ProductItem
 */
 ProductItem = function () { 
	this.Title = "";

	this.BodyHtml = "";

	this.Vendor = "";

	this.ProductType = "";

	this.Variants = new Array();

}

/**
 * bNesis unified Coutry item for services unified API call 
 * @typedef {Object} CountryItem
 */
 CountryItem = function () { 
	/**
	 * Country data block (record) unical Id
	 * @type {string}
	 */
	this.Id = "";

	/**
	 * The Name of the Country
	 * @type {string}
	 */
	this.Name = "";

	/**
	 * Coutry Code
	 * https://en.wikipedia.org/wiki/Country_code
	 * ISO
	 * @type {string}
	 */
	this.Code = "";

	/**
	 * https://en.wikipedia.org/wiki/List_of_countries_by_tax_rates
	 * @type {string}
	 */
	this.Tax = "";

}

/**
 * bNesis balance class for unified API 
 * @typedef {Object} Balance
 */
 Balance = function () { 
	/**
	 * currency type flag
	 * @type {string}
	 */
	this.Currency = "";

	/**
	 * Currency ammount value
	 * @type {Double}
	 */
	this.Amount = new Double();

}

/**
 * bNesis balance result unified class 
 * @typedef {Object} RetrieveBalanceUnified
 */
 RetrieveBalanceUnified = function () { 
	/**
	 * this record Id - define data block at service
	 * @type {string}
	 */
	this.Id = "";

	/**
	 * Amount value
	 * @type {Double}
	 */
	this.Amount = new Double();

	/**
	 * Currency type flag
	 * @type {string}
	 */
	this.Currency = "";

	/**
	 * data block (record) creation date time
	 * @type {string}
	 */
	this.Created = "";

}

 PaymentsArchiveData = function () { 
	this.PaymentId = 0;

	this.Amount = new Double();

	this.Currency = "";

	this.CreateDate = 0;

	this.TransactionId = 0;

}

 PaymentsArchive = function () { 
	this.data = new Array();

}

